### PR TITLE
[ads] Fix new tab page video ad media 25 attribution

### PR DESCRIPTION
**Pull request by aseren**
Resolves https://github.com/brave/brave-browser/issues/40090

## Submitter Checklist:

- [x] I confirm that no [security/privacy review is needed](https://github.com/brave/brave-browser/wiki/Security-reviews) and no other type of reviews are needed, or that I have [requested](https://github.com/brave/reviews/issues/new/choose) them
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally:
  * `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests` [wiki](https://github.com/brave/brave-browser/wiki/Tests)
  * `npm run presubmit` [wiki](https://github.com/brave/brave-browser/wiki/Presubmit-checks), `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### Prerequisites to install staging components

New Tab Takeover video ads are available in staging component updater for iOS.
Steps to setup staging component updater on iOS:
* Fresh install
* Launch browser
* Toggle `Use Staging CRX components` in `Settings -> BraveCore Switches`
* Close the browser
* Launch Brave browser
* Join Brave Rewards
* Staging components are downloaded and applied in a minute after start

### Test case
* Get iPhone as a device
* Do Brave fresh install
* Do prerequisites to install staging components
* Rotate iPhone to landscape orientation
* Open New Tab Pages until a tab with New Tab Takeover ad is opened
* Rotate iPhone to portrait orientation
* Check that the NTT video and Play button become visible
* Tap Play button
* Wait until video finishes playing
* Check that Brave Ads media play event was fired
   EXPECTATION: There should be line in Brave Ads logs:
	`[ads] Started playing new tab page video ad with placement id % and creative instance id %`
* Check that Brave Ads media 25 event was fired
   EXPECTATION: There should be line in Brave Ads logs:
	`[ads] Played 25% of new tab page video ad with placement id % and creative instance id %`
* Check that Brave Ads media 100 event was fired
   EXPECTATION: There should be line in Brave Ads logs:
	`[ads] Played 100% of new tab page video ad with placement id % and creative instance id %`

### CODE REVIEW

**Review by tmancey — APPROVED** [review]
LGTM++

File: ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift [hunk 1]
```diff
@@ -133,9 +103,16 @@ class NewTabPageVideoPlayer {
     return !wasPlaying
   }
 
-  func maybeStartAutoplay() {
-    if !didStartAutoplay {
-      loadTrackParams(backgroundVideoPath)
+  func maybeLoadVideoAssetAndStartAutoplay(shouldStartAutoplay: Bool) {
```
Comment (aseren) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift:106]: we found `IfNeeded` is used more then `maybe` prefix for iOS code.
@kylehickinson is it better to use `loadVideoAssetIfNeededAndAutoplay` naming here?

Comment (aseren) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift]: changed to `loadAndAutoplayVideoAssetIfNeeded `

File: ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift [hunk 2]
```diff
@@ -26,11 +26,12 @@ class NewTabPageVideoPlayer {
...
   private var didFinishAutoplay = false
-  private var timeObserver: Any?
-  private var playerObserver: NSKeyValueObservation?
+
+  private var media25TimeObserver: Any?
+  private var media100TimeObserver: Any?
+  private var autoplayFinishedObserver: NSKeyValueObservation?
 
   private let kMaxAutoplayDurationInSeconds = 6.0
   private let kMedia25Percent = 0.25
```
Comment (soner-yuksel) [inline ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift:37]: I dont think this value is not used anymore

**Review by soner-yuksel — APPROVED** [review]
Only a nit, rest seems good

File: ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift (hunk 2 above)
Comment (aseren) [reply ios/brave-ios/Sources/Brave/Frontend/Browser/New Tab Page/Backgrounds/NewTabPageVideoPlayer.swift]: removed

**Comment by btlechowski** [comment]
Verified on 1.70.49 on iPhone 13 Pro Max

Verified test plan from https://github.com/brave/brave-core/pull/24955

image|image

Verified served and view events were fired:
```
info	13:45:35.553381+0200	Client	[ads] Served new tab page ad impression with placement id ACA3FBBB-60DC-495E-8729-700601F9ACB7 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998
info	13:45:35.565145+0200	Client	[ads] Viewed new tab page ad impression with placement id ACA3FBBB-60DC-495E-8729-700601F9ACB7 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998
```

Verified Brave Ads media play event was fired
```
info	13:46:52.481251+0200	Client	[ads] Started playing new tab page video ad with placement id ACA3FBBB-60DC-495E-8729-700601F9ACB7 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998
info	13:46:52.490411+0200	Client	[ads] Confirming media_play for new_tab_page_ad with transaction id 1d7084d7-e206-4a88-b57d-3131ae653289 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998

```

Verified Brave Ads media 25 event was fired
```
info	13:46:56.415329+0200	Client	[ads] Played 25% of new tab page video ad with placement id ACA3FBBB-60DC-495E-8729-700601F9ACB7 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998
info	13:46:56.419528+0200	Client	[ads] Confirming media_25 for new_tab_page_ad with transaction id 28cd7f35-9998-48b9-815a-f74086f50c87 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998

```

Verified Brave Ads media 100 event was fired
```
info	13:47:07.653212+0200	Client	[ads] Played 100% of new tab page video ad with placement id ACA3FBBB-60DC-495E-8729-700601F9ACB7 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998
info	13:47:07.655515+0200	Client	[ads] Confirming media_100 for new_tab_page_ad with transaction id cf33a730-24ed-4add-b6fc-8e645fb21fc0 and creative instance id 69b9626e-ca72-484a-8086-54f88bc15998

```

**Pull request merged**